README.md
# spreadstream
[![Build Status](https://travis-ci.org/nodys/spreadstream.svg?branch=master)](https://travis-ci.org/nodys/spreadstream) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com) [![Clussh on npm](https://img.shields.io/npm/v/spreadstream.svg)](https://www.npmjs.com/package/spreadstream)

Pipe data from and to google spreadsheet

---

<p align="center">
  <img src="https://nodys.github.io/spreadstream/spreadstream.png" alt="clussh">
</p>

Read
```sh
spreadstream > data.csv
spreadstream --json > data.ndjson
```

Write
```sh
cat data.csv | spreadstream
cat data.ndjson | spreadstream --json
```

---

##  Install

```sh
npm install -g spreadstream
```

Then:

1. [Create a Google authentication token](#google-authentication-token) and make sure the account (or the service account) can access the spreadsheets you plan to use.
2. Create a [rc file](#rc-file) with your authentication token


## Configuration

### Google authentication token
You need to create a Google authentication token for the Google Sheet Api: either a [service account](#create-a-service-account) or a [OAuth 2 token](#create-oauth2-token) to use your own account.

Once created, put your credentials in a [rc file](#rc-file) as described below.


### Rc file
The rc file must contain the `credential` key with the google authentication token created previously. You can add any other spreadstream options (see `spreadstream --help`). The location of the rc file depend on your needs: either at a [standard rc file path](https://www.npmjs.com/package/rc) or specified using the `--settings` option.

*You can too try the new (alpha) spreadstream config generator which supports Oauth2 authentication: `spreadstream init`*

**Exemple:**

```js
// .spreadstreamrc
{
  // Your google authentication token created previously:
  "credential": {
    "type": "service_account",
    "project_id": "xxxxx",
    "private_key_id": "xxxxx",
    "private_key": "xxxxx",
    "client_email": "xxxxx@xxxxx",
    "client_id": "xxxxx",
    "auth_uri": "https://accounts.google.com/o/oauth2/auth",
    "token_uri": "https://accounts.google.com/o/oauth2/token",
    "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
    "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/xxxxx"
  },
  // You can set default value for any spreadstream command line options...
  "id": "spreadsheet id", // The document id from the spreadsheet url
  "sheet": "My Sheet",    // The sheet title in your document
  "json": true
  // ...
}
```

## Usage

The examples below depends on the availability of a [rc file](#rc-file) containing a valid `credential`, a spreadsheet document `id` and a `sheet` title.

### Write data to Google Spreadsheet

```sh
# Pipe csv to spreadstream (append new rows to the sheet)
cat mydocument.csv | spreadstream

# Or read a csv file
spreadstream --input mydocument.csv

# Choose the document and the sheet (or use configuration file):
cat mydocument.csv | spreadstream --id="ya29.GlsiBTHclgwXhCs3dJZHp" --sheet "My Sheet"

# Clear sheet first (replace)
cat mydocument.csv | spreadstream --replace

# Pipe line delimited json instead of csv
cat mydocument.ndjson | spreadstream --json
```

### Read data from Google Spreadsheet

```sh
# Read sheet
spreadstream

# Choose the document and the sheet (or use configuration file):
spreadstream --id="ya29.GlsiBTHclgwXhCs3dJZHp" --sheet "My Sheet"

# Limit reading range with A1 notation
# https://developers.google.com/sheets/api/guides/concepts#a1_notation
spreadstream --range="A:C"
spreadstream --range="A1:C4"
spreadstream --range="1:4"
spreadstream --range="My Sheet!1:4" # Override sheet

# Change csv output (same options than for input):
spreadstream --csv-separator ";"

# output as line delimited json instead of csv
spreadstream --json

# Write output to a file
spreadstream > myfile.csv
spreadstream --output myfile.csv
```

### Options

See too `spreadstream --help` for detailed command line usage, options and default values.

The [API](#api) use the camelCase version for the dashed options names. Every option can be set in the [rc file](#rc-file) using either camelCase or kebab-case format.

- `--id`
   Identifier of the spreadsheet document
   <small>The spreadsheet document id is the long unique identifier in the URL of the document looking like `Dh9CsT4eXiTeKQLWZLpM..`</small>
- `--sheet`
   Title of the sheet in the document
- `--replace`
   Write in overwrite mode: replace the content of the sheet (the default behavior is to append new rows at the end)
- `--verbose`
  Be verbose about what is done (on stderr)
- `--value-input`
  Determines how input data should be interpreted (default: `USER_ENTERED`) ([more](https://developers.google.com/sheets/api/reference/rest/v4/ValueInputOption))
- `--major-dimension`
  Indicates which dimension read operation should apply to (default: `ROWS`) ([more](https://developers.google.com/sheets/api/reference/rest/v4/Dimension))
- `--value-render`
  Determines how values should be rendered in the the output while reading (default: `FORMATTED_VALUE`) ([more](https://developers.google.com/sheets/api/reference/rest/v4/valueRender))
- `--date-time-render`
   Determines how dates should be rendered in the the while reading (default: `SERIAL_NUMBER`) ([more](https://developers.google.com/sheets/api/reference/rest/v4/dateTimeRender))
- `--max-buffer`
   Buffer max size before flushing to spreadsheet (default: `1000`).
   How many row of data should be sent at once to the spreadsheet while writing.
   A lower value would negatively impact speed and API usage limits, but it will produce atomic update of your document when used with light and slow stream of data (eg. a line-delimited-json log producer).
- `--range`
   Fore reading: The A1 notation of the values to retrieve. Default is to select the whole sheet.
   Exemples: `A1:D3` a 4x3 range, `A:D` the four first columns, `12:30` for lines from 12 to 30
- `--csv-separator`
   Csv separator (both for the parser and the writer).
   Default to auto-detect and `,`
- `--csv-quote`
   Csv quote (both for the parser and the writer)
   Default to auto-detect and `"`
- `--csv-escape`
   Csv quote escaping (both for the parser and the writer)
   Default to auto-detect and `""`
- `--csv-newline`
   Csv new line character
   Default to auto-detect and `\n`
- `--read-headers` (default: true)
  The first chunk in the input feed should be used as headers (prefix with `--no-` to disable)
  If false:
  - When reading from json, this option has no effect.
  - When reading from csv, the first line is not interpreted as headers (and default headers are generated)
  - When reading from spreadsheet, the first line is not interpreted as headers (and default headers are generated)
- `--write-headers` (default: true)
  The first chunk in the output feed should include headers
  If false:
  - When writing to json, this option has no effect.
  - When writing to csv the headers are omitted
  - When writing to spreadsheet the headers are omitted
- `--noheaders` (default: false)
  Alias for `--no-read-headers` and `--no-write-headers`
- `--json`
   Use [new line delimited json](http://ndjson.org/) parser and writer instead of csv as input and output.
- `--classic-json`
   Input / output format should use classic json serializer and parser instead of line delimited json (read and produce a json array)
- `--input`
   Set input file. Default is to read from stdin outside of a tty environment.
   Reading from stdin can be forced by setting this option to `-`.
- `--output`
   Set output file. Default is to write to stdout.
   Writing to stdout can be forced by setting this option to `-`.
- `completion`
   Generate bash completion code (`spreadstream completion >> $HOME/.bashrc`)

## Api

```js
const spreadstream = require('spreadstream')

const config = {
  // See below about creating a service account (required)
  credentials: require('./google-credentials.json')  

  // Spreadsheet id (see document's url) shared with the service account (required)
  id: '1jLPcDv0UaYIDh9CsT4eXiMeKQLWZLpMyAA7FjkO7Z3X',

  // The sheet title (required)
  sheet: 'My Sheet',

  // Clear sheet before adding value (default: false)
  // If true, every values in the sheet will be removed first.
  // The default behavior (replace: false) is to append rows at the bottom of
  // the sheet.
  // Headers are pushed only when replace is true or when the sheet is new.
  replace: false,

  // Input value option:
  // - USER_ENTERED: The values will be parsed as if the user typed them into the UI (the default)
  // - RAW: The values will be stored as-is.
  // See https://developers.google.com/sheets/api/reference/rest/v4/ValueInputOption
  valueInputOption: spreadstream.USER_ENTERED,

  // How many row must be keept in the stream buffer before flushing
  // data to the document (default: 5000)
  maxBuffer: 5000,

  // Verbose mode (default: false)
  verbose: false
}

// Create a stream
const stream = spreadstream(config)

// Pipe an object stream or write directly to the stream.
// The spreadstream stream supports to kind of object stream: Array or Object.
// Note that you can not push a mixed stream of array and object.

// Stream of Array:
// Each array is a row in the sheet. The first row will be interpreted has
// headers (unless readHeaders if false)
const stream1 = spreadstream(config)
stream1.write(['foo', 'bar'])
stream1.write(['4', '2'])
stream1.write(['7', '10'])
stream1.end()


// Stream of Object
const stream2 = spreadstream(config)
stream2.write({ foo: 4, bar: 2 })
stream2.write({ foo: 7, bar: 10 })
stream2.end()


// Pipe data from stdin (eg. with ndjson):
process.stdin
  .pipe(require('ndjson').parse())
  .pipe(spreadstream(config))

// Read a document
spreadstream.readDocument(config).then(values => console.log(values))

```

## Create a service account

1. Go to the [Google Developers Console](https://console.developers.google.com/project)
2. Select your project or create a new one (and then select it)
3. Enable the Google Sheets API for your project
   - Search for "sheet"
   - In the sidebar on the left, expand __APIs & auth__ > __APIs__
   - Click on "[Google Sheets API](https://console.developers.google.com/apis/api/sheets.googleapis.com/overview)"
   - Click the blue "Enable API" button
4. Create a service account for your project
   - In the sidebar on the left, expand __APIs & auth__ > __Credentials__
   - Click blue "Create credentials" button
   - Select the "Create service account key" option
   - Select "New service account"
   - Keep the "JSON" key type option selected
   - Click blue "Create" button and complete the form
   - Your JSON key file is generated and downloaded to your machine
     (__it is the only copy!__)
   - Note your service account's *email address* (also available in the JSON
     key file)
5. Add the content of the JSON file in your spreadstream rc file (eg. `./.spreadstreamrc` or `~/.config/spreadstream/config`):
   ```json
   {
     "credential": {
        "type": "service_account",
        "project_id": "xxxxx",
        "private_key_id": "xxxxx",
        "private_key": "xxxxx",
        "client_email": "xxxxx@xxxxx",
        "client_id": "xxxxx",
        "auth_uri": "https://accounts.google.com/o/oauth2/auth",
        "token_uri": "https://accounts.google.com/o/oauth2/token",
        "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
        "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/xxxxx"
      }
   }
   ```
6. Share the doc (or docs) with your service account using the email
   noted above

## Create Oauth2 token

1. Go to the [Google Developers Console](https://console.developers.google.com/project)
2. Select your project or create a new one (and then select it)
3. Enable the Google Sheets API for your project
   - Search for "sheet"
   - In the sidebar on the left, expand __APIs & auth__ > __APIs__
   - Click on "[Google Sheets API](https://console.developers.google.com/apis/api/sheets.googleapis.com/overview)"
   - Click the blue "Enable API" button
4. Create a Oauth credential
   - In the sidebar on the left, expand __APIs & auth__ > __Credentials__
   - Click blue "Create credentials" button
   - Select the "OAuth 2.0 client IDs" option
   - Select "Other" and give a name
   - Click blue "Create" button
   - Copy the client id (`CLIENT_ID` below) and the secret key (`CLIENT_SECRET` below)
5. Edit you spreadstream rc file (eg. `~/.config/spreadstream/config`):
   ```json
   {
     "type": "oauth2",
     "client_id": "CLIENT_ID",
     "client_secret": "CLIENT_SECRET",
     "tokens": { }
   }
   ```
6. Complete the


*(credit: [node-google-spreadsheet]( https://github.com/theoephraim/node-google-spreadsheet/blob/master/README.md#service-account-recommended-method))*

---

License: [MIT](./LICENSE) - Novadiscovery


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [4.0.4](https://github.com/nodys/spreadstream/compare/v4.0.3...v4.0.4) (2021-03-05)


### Bug Fixes

* bump dependencies ([b5cf64d](https://github.com/nodys/spreadstream/commit/b5cf64d618c37fe73e0c2fd7f0c10a2151247f21))

### [4.0.3](https://github.com/nodys/spreadstream/compare/v4.0.2...v4.0.3) (2020-06-05)


### Bug Fixes

* dependency upgrade ([3770dff](https://github.com/nodys/spreadstream/commit/3770dffbf9a50acd93746d4d1da1852dde8f0fbc))

### [4.0.2](https://github.com/nodys/spreadstream/compare/v4.0.1...v4.0.2) (2020-06-05)


### Bug Fixes

* enforce output does not work with stdout ([5658bcd](https://github.com/nodys/spreadstream/commit/5658bcd3cf8eacdb0657926231471e5c64d666ba))

### [4.0.1](https://github.com/nodys/spreadstream/compare/v3.3.3...v4.0.1) (2020-04-15)

## [4.0.0](https://github.com/nodys/spreadstream/compare/v3.3.3...v4.0.0) (2020-04-15)

<a name="3.3.3"></a>
## [3.3.3](https://github.com/nodys/spreadstream/compare/v3.3.2...v3.3.3) (2019-05-08)


### Bug Fixes

* Upgrade dependencies ([89a8741](https://github.com/nodys/spreadstream/commit/89a8741))



<a name="3.3.2"></a>
## [3.3.2](https://github.com/nodys/spreadstream/compare/v3.3.1...v3.3.2) (2018-12-07)


### Bug Fixes

* Serialize array from json ([fddbb42](https://github.com/nodys/spreadstream/commit/fddbb42))



<a name="3.3.1"></a>
## [3.3.1](https://github.com/nodys/spreadstream/compare/v3.3.0...v3.3.1) (2018-08-20)


### Bug Fixes

* Invalid config path in init ([194de8b](https://github.com/nodys/spreadstream/commit/194de8b))



<a name="3.3.0"></a>
# [3.3.0](https://github.com/nodys/spreadstream/compare/v3.2.1...v3.3.0) (2018-08-03)


### Features

* Enhance initializer and fix dependencies ([400ea15](https://github.com/nodys/spreadstream/commit/400ea15))



<a name="3.2.1"></a>
## [3.2.1](https://github.com/nodys/spreadstream/compare/v3.2.0...v3.2.1) (2018-04-30)


### Bug Fixes

* Bug introduced in last version with lsjson ([e2fa8e1](https://github.com/nodys/spreadstream/commit/e2fa8e1))



<a name="3.2.0"></a>
# [3.2.0](https://github.com/nodys/spreadstream/compare/v3.1.3...v3.2.0) (2018-04-09)


### Features

* Add experimental transform for output ([39d00e7](https://github.com/nodys/spreadstream/commit/39d00e7))



<a name="3.1.3"></a>
## [3.1.3](https://github.com/nodys/spreadstream/compare/v3.1.2...v3.1.3) (2018-04-03)


### Bug Fixes

* Hnadle properly unformatted values ([9c86c4d](https://github.com/nodys/spreadstream/commit/9c86c4d))


### Features

* **json:** csvFieldToJson can now be configured ([c26753c](https://github.com/nodys/spreadstream/commit/c26753c))



<a name="3.1.2"></a>
## [3.1.2](https://github.com/nodys/spreadstream/compare/v3.1.1...v3.1.2) (2018-03-29)


### Bug Fixes

* **json:** Empty cell should be converted to null ([930fb1a](https://github.com/nodys/spreadstream/commit/930fb1a))



<a name="3.1.1"></a>
## [3.1.1](https://github.com/nodys/spreadstream/compare/v3.1.0...v3.1.1) (2018-03-29)


### Bug Fixes

* **json:** Classic json stream should support empty stream ([8b1afc3](https://github.com/nodys/spreadstream/commit/8b1afc3))



<a name="3.1.0"></a>
# [3.1.0](https://github.com/nodys/spreadstream/compare/v3.0.1...v3.1.0) (2018-03-27)


### Features

* **stream:** Expose internal classic json stream ([7194d90](https://github.com/nodys/spreadstream/commit/7194d90))



<a name="3.0.1"></a>
## [3.0.1](https://github.com/nodys/spreadstream/compare/v3.0.0...v3.0.1) (2018-03-26)


### Bug Fixes

* **json:** Invalid boolean parser ([aaa99bd](https://github.com/nodys/spreadstream/commit/aaa99bd))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/nodys/spreadstream/compare/v2.2.0...v3.0.0) (2018-03-22)


### Bug Fixes

* **option:** Normalize json-classic option ([2784fce](https://github.com/nodys/spreadstream/commit/2784fce))
* **option:** Normalize value render option and value input option names ([995aa69](https://github.com/nodys/spreadstream/commit/995aa69))
* **option:** Read properly valueInputOption ([36591f4](https://github.com/nodys/spreadstream/commit/36591f4))
* **test:** Add the graceful option ([5621a01](https://github.com/nodys/spreadstream/commit/5621a01))


### Features

* **error:** Throw an error if the source sheet does not exists fix [#1](https://github.com/nodys/spreadstream/issues/1) ([5c7e098](https://github.com/nodys/spreadstream/commit/5c7e098))
* **json:** Add value parser for json values for boolean and number ([fab795f](https://github.com/nodys/spreadstream/commit/fab795f))



<a name="2.2.0"></a>
# [2.2.0](https://github.com/nodys/spreadstream/compare/v2.1.0...v2.2.0) (2018-03-22)


### Bug Fixes

* **init:** Fix typo in rc file output ([4730f37](https://github.com/nodys/spreadstream/commit/4730f37))


### Features

* **doc:** Add a notice about the init tool ([459fe50](https://github.com/nodys/spreadstream/commit/459fe50))
* **stream:** Add new stream output --classic-json ([f76aa85](https://github.com/nodys/spreadstream/commit/f76aa85))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/nodys/spreadstream/compare/v2.0.0...v2.1.0) (2018-03-07)


### Bug Fixes

* **standard:** Fix style ([e469b2d](https://github.com/nodys/spreadstream/commit/e469b2d))


### Features

* Add spreadstream init tool ([3287401](https://github.com/nodys/spreadstream/commit/3287401))



<a name="2.0.0"></a>
# [2.0.0](https://github.com/nodys/spreadstream/compare/v1.2.0...v2.0.0) (2018-03-07)


### Features

* **headers:** Add new headers options ([8dc1562](https://github.com/nodys/spreadstream/commit/8dc1562))


### BREAKING CHANGES

* **headers:** the `--headers` option has been removed



<a name="1.2.0"></a>
# [1.2.0](https://github.com/nodys/spreadstream/compare/v1.1.0...v1.2.0) (2018-03-07)


### Features


test/utils/mock-google-sheets.js
module.exports = makeMock

function makeMock () {
  const mocked = {
    get: [],
    batchUpdate: [],
    values: {
      get: [],
      clear: [],
      append: []
    }
  }
  const mock = {
    mocked,
    spreadsheets: {
      get (query, cb) {
        const { spreadsheetId } = query
        mocked.get.push(query)
        cb(null, {
          data: {
            spreadsheetId,
            sheets: [ makeMock.makeSheet() ]
          }
        })
      },

      batchUpdate (query, cb) {
        mocked.batchUpdate.push(query)
        cb(null, { data: { replies: query.resource.requests.map(q => {
          const key = Object.keys(q)[0]
          const param = q[key]
          return param
        })}})
      },

      values: {
        get: (params, cb) => {
          mocked.values.get.push(params)
          cb(null, { data: { values: [ [ 'foo', 'bar' ], [ '24', '-10E-3' ], [ 'TRUE', 'FALSE' ], [ '', '0' ] ] } })
        },
        clear: (data, cb) => {
          mocked.values.clear.push(data)
          cb()
        },
        append: (data, cb) => {
          mocked.values.append.push(data)
          cb()
        }
      }
    }
  }
  return mock
}

makeMock.makeSheet = ({ title = 'test', rowCount = 2, columnCount = 2 } = {}) => {
  return { properties: { title: 'test', gridProperties: { rowCount, columnCount } } }
}


lib/enums.js

/**
 * Type of objects from input stream
 * @type {Object}
 */
exports.streamType = {
  ARRAY: 'ARRAY',
  OBJECT: 'OBJECT'
}
/**
 * Type of objects from input stream
 * @type {Object}
 */
exports.mode = {
  READING: 'READING',
  WRITING: 'WRITING'
}

/**
 * Determines how input data should be interpreted
 * @type {Object}
 * @see https://developers.google.com/sheets/api/reference/rest/v4/ValueInputOption
 */
exports.valueInputOption = {
  USER_ENTERED: 'USER_ENTERED',
  RAW: 'RAW'
}

/**
 * Indicates which dimension read operation should apply to
 * @type {Object}
 * @see https://developers.google.com/sheets/api/reference/rest/v4/Dimension
 */
exports.majorDimension = {
  COLUMNS: 'COLUMNS',
  ROWS: 'ROWS'
}

/**
 * Determines how values should be rendered in the the output.
 * @type {Object}
 * @see https://developers.google.com/sheets/api/reference/rest/v4/valueRender#ENUM_VALUES.FORMATTED_VALUE
 */
exports.valueRenderOption = {
  FORMATTED_VALUE: 'FORMATTED_VALUE',
  UNFORMATTED_VALUE: 'UNFORMATTED_VALUE',
  FORMULA: 'FORMULA'
}

/**
 * Determines how dates should be rendered in the the output.
 * @type {Object}
 * @see https://developers.google.com/sheets/api/reference/rest/v4/dateTimeRender
 */
exports.dateTimeRender = {
  SERIAL_NUMBER: 'SERIAL_NUMBER',
  FORMATTED_STRING: 'FORMATTED_STRING'
}


lib/streams.js
const miss = require('mississippi')

/**
 * Create a classic json output stream (write a json array)
 * @return {stream.Passthrough}
 */
exports.classicJsonOutputStream = function classicJsonOutputStream (replacer = null, space = 2) {
  let first = true
  return miss.through.obj(function (data, enc, next) {
    let chunk = ''
    if (first) {
      chunk = '[\n'
    }
    if (!first) {
      chunk += ',\n'
    }
    chunk += JSON.stringify(data, replacer, space).split('\n').map(r => `  ${r}`).join('\n')
    if (first) {
      first = false
    }
    this.push(chunk)
    next()
  }, function (done) {
    if (first) {
      this.push('[]\n')
    } else {
      this.push('\n]\n')
    }
    done()
  })
}

/**
 * Create a classic json input stream (read a json array)
 * @return {stream.Passthrough}
 */
exports.classicJsonInputStream = function classicJsonInputStream () {
  let buffer = []
  return miss.through.obj(function (chunk, enc, next) {
    buffer.push(chunk.toString())
    next()
  }, function (done) {
    let data
    try {
      data = JSON.parse(buffer.join(''))
    } catch (error) {
      return done(new Error(`Invalid input type: unable to parse (original message: ${error.message})`))
    }
    if (!Array.isArray(data)) {
      return done(new Error('Invalid input type: array expected with classic json stream'))
    }
    for (let row of data) {
      this.push(row)
    }
    done()
  })
}


lib/google-wrapper.js
const { google } = require('googleapis')
const { GoogleAuth, OAuth2Client } = require('google-auth-library')

/**
 * Sheet api
 * @type {Object}
 */
exports.sheets = google.sheets('v4')

/**
 * Authenticate service account
 * @param  {Object} credential Service account credential
 * @return {Promise<Object>} Google auth instance
 */
/* istanbul ignore next reason: Mocked during test */
exports.authenticate = async function (credential) {
  switch (credential.type) {
    case 'service_account':
      const googleAuth = new GoogleAuth()
      const auth = googleAuth.fromJSON(credential)
      auth.scopes = ['https://www.googleapis.com/auth/spreadsheets']
      await auth.authorize()
      return auth
    case 'oauth2':
      const oAuth2Client = new OAuth2Client(
        credential.client_id,
        credential.client_secret,
        'urn:ietf:wg:oauth:2.0:oob'
      )
      oAuth2Client.setCredentials(credential.tokens)
      return oAuth2Client
    default:
      throw new Error('Unregistered credential type')
  }
}


lib/spreadstream.js
const { fromCallback } = require('bluebird')
const miss = require('mississippi')
const chalk = require('chalk')
const enums = require('./enums')
const google = require('./google-wrapper')
const streams = require('./streams')

module.exports = spreadstream

/**
 * Create spreadstream writable stream
 * @param  {Object} config  Configuration
 * @param  {String} config.id  spreadsheet id
 * @param  {String} config.sheet  spreadsheet sheet name
 * @param  {Boolean} [config.replace] Replace current sheet value (default: false, append)
 * @param  {Number} [config.maxBuffer] Max stream buffer size before flushing to document
 * @param  {String} [config.valueInputOption] See google sheet api for input option
 * @param  {String} [config.verbose] Be verbose about progression
 * @param  {Boolean} [config.noheaders=false] Force config.writeHeaders and config.writeHeaders to false
 * @param  {Boolean} [config.writeHeaders=true] Write headers (default: true)
 * @param  {Boolean} [config.readHeaders=true] Read headers from first source chunk (default: true)
 * @return {Writeable} Writeable object stream
 * @see https://developers.google.com/sheets/api/guides/concepts
 */
function spreadstream (config) {
  config = spreadstream._normalizeConfig(config)

  // Check required configs
  if (!config.id) throw new Error('Missing required config: spreadsheet id')
  if (!config.sheet) throw new Error('Missing required config: sheet title')

  // Verbose logger utility (or noop)
  const log = config.verbose /* istanbul ignore next */
    ? m => console.error(chalk`[{green ${(new Date()).toISOString()}}] ${m}`)
    : m => m

  // Total row counter
  let total = 0

  // Stream buffer
  let buffer = []

  // Google spreadsheet initialized document
  let doc

  // Headers
  let headers

  // Type of stream (Array or Object)
  let streamType

  async function flush () {
    if (!buffer.length) return

    let values = []

    if (!doc) {
      log(`Load document and sheet`)
      doc = await spreadstream.writableDocument(config)
    }

    // If first flush...
    if (flush.count === 0) {
      // Remove first line (headers)
      let firstRow = buffer[0]

      // Store type of stream
      streamType = Array.isArray(firstRow) ? enums.streamType.ARRAY : enums.streamType.OBJECT

      // Store headers
      if (config.readHeaders) {
        if (streamType === enums.streamType.ARRAY) {
          headers = [...firstRow]
          buffer.shift()
        } else {
          headers = Object.keys(firstRow)
        }
      } else {
        if (streamType === enums.streamType.ARRAY) {
          headers = Array(firstRow.length).fill(0).map((_, index) => toColumnName(index + 1))
        } else {
          // readHeaders has no effect on object stream
          headers = Object.keys(firstRow)
        }
      }

      // Add headers (first flush, with write header on and we are not in append mode)
      if (doc.addHeaders && config.writeHeaders) {
        log('Add headers')
        values.push(headers)
      }
    }

    // Push rows
    for (let row of buffer) {
      if (streamType === enums.streamType.OBJECT) {
        values.push(headers.map(key => row[key]))
      } else {
        values.push(row)
      }
    }

    total += values.length

    log(`Flush ${values.length} rows`)

    await doc.push(values)

    buffer = []
    flush.count++
  }

  // Count flush calls (and trac first flush to add headers)
  flush.count = 0

  return miss.to.obj(function (row, _, done) {
    buffer.push(row)
    if (buffer.length >= config.maxBuffer) {
      flush().then(() => done(), done)
    } else {
      done()
    }
  }, async function (done) {
    try {
      await flush()
      log(`Success (${total} rows inserted)`)
      done()
    } catch (error) {
      done(error)
    }
  })
}

// Expose api
spreadstream.google = google
spreadstream.enums = enums
spreadstream.streams = streams

/**
 * Retrurn proper config --
 * @param  {object} _config
 * @return {object}
 */
spreadstream._normalizeConfig = function (_config) {
  const config = {
    writeHeaders: true,
    readHeaders: true,
    ..._config
  }

  if (config.noheaders === true) {
    config.writeHeaders = false
    config.readHeaders = false
  }

  return config
}

/**
 * Load spreadsheet document (for writing)
 * @param  {Object} config  Configuration
 * @param  {String} config.id  spreadsheet id
 * @param  {String} config.sheet  spreadsheet sheet name
 * @param  {String} config.valueInputOption  See google sheet valueInputOption
 * @param  {Boolean} [config.replace] Replace current sheet value (default: false, append)
 * @return {Promise} Returns an object with
 *   - {function} push({Array[] value) - Push values
 *   - {boolean} addHeaders - Headers row expected in first push (replace is true or sheet is new)
 *   - {object} spreadsheet - Spreadsheet instance
 *   - {object} sheet - Sheet instance
 *
 * @see https://developers.google.com/sheets/api/guides/concepts
 */
spreadstream.writableDocument = async function (config) {
  config = spreadstream._normalizeConfig(config)

  // Check config
  if (!config.id) throw new Error('Missing required config: spreadsheet id')
  if (!config.sheet) throw new Error('Missing required config: sheet title')

  const spreadsheetId = config.id
  const sheetTitle = config.sheet
  const valueInputOption = config.valueInputOption || enums.valueInputOption.USER_ENTERED
  const auth = config._auth || await spreadstream.google.authenticate(config.credential)
  const sheetsApi = spreadstream.google.sheets

  // Should we replace the previous values with new one ?
  let replace = Boolean(config.replace)

  // Should we keep header line
  let addHeaders = replace

  // Load spreadsheet document
  let spreadsheet = await fromCallback(cb => sheetsApi.spreadsheets.get({
    auth,
    spreadsheetId,
    ranges: []
  }, cb))

  // Load the sheet (from title)
  let sheet = spreadsheet.data.sheets.find(s => s.properties.title === sheetTitle)
  // If the sheet does not exists let create one
  if (!sheet) {
    addHeaders = true
    let addSheet = await fromCallback(cb => sheetsApi.spreadsheets.batchUpdate({
      auth,
      spreadsheetId: spreadsheet.data.spreadsheetId,
      resource: {
        requests: [{
          addSheet: {
            properties: { title: sheetTitle, gridProperties: { rowCount: 1, columnCount: 1 } }
          }
        }]
      }
    }, cb))
    sheet = addSheet.data.replies[0].addSheet
  }

  // Clear current sheet
  if (replace) {
    let range = `A1:${toColumnName(sheet.properties.gridProperties.columnCount)}${sheet.properties.gridProperties.rowCount}`
    await fromCallback(cb => sheetsApi.spreadsheets.values.clear({
      auth,
      spreadsheetId,
      range: `'${sheetTitle}'!${range}`,
      resource: {}
    }, cb))
  }

  // Push data to document (append)
  async function push (rows) {
    // Normalize Array values:
    rows = rows.map(row => {
      return row.map(value => Array.isArray(value) ? value.join('\n') : value)
    })

    // Push values to spreadsheet
    return fromCallback(cb => sheetsApi.spreadsheets.values.append({
      auth,
      spreadsheetId,
      range: `'${sheetTitle}'!A1`,
      valueInputOption,
      resource: { values: rows }
    }, cb))
  }

  return {
    addHeaders,
    spreadsheet,
    sheet,
    push
  }
}

/**
 * Read document
 * @param  {Object} config configuration
 * @param  {String} config.id  spreadsheet id
 * @param  {String} config.sheet  spreadsheet sheet name
 * @param  {Boolean} [config.readHeaders=true] Read headers from first source chunk (default: true)
 * @return {Array} Results
 */
spreadstream.readDocument = async function (config) {
  config = spreadstream._normalizeConfig(config)

  // Check required configs
  if (!config.id) throw new Error('Missing required config: spreadsheet id')
  if (!config.sheet) throw new Error('Missing required config: sheet title')

  const spreadsheetId = config.id
  const sheetTitle = config.sheet
  const auth = config._auth || await spreadstream.google.authenticate(config.credential)
  const sheetsApi = spreadstream.google.sheets
  const valueRenderOption = config.valueRenderOption || enums.valueRenderOption.FORMATTED_VALUE
  const dateTimeRenderOption = config.dateTimeRender || enums.dateTimeRender.SERIAL_NUMBER
  const majorDimension = config.majorDimension
  const jsonRaw = config.jsonRaw
  const csvFieldToJson = config.csvFieldToJson || spreadstream.csvFieldToJson

  // Verbose logger utility (or noop)
  const log = config.verbose /* istanbul ignore next */
    ? m => console.error(chalk`[{green ${(new Date()).toISOString()}}] ${m}`)
    : m => m

  let range = config.range
    ? `'${sheetTitle}'!${config.range}`
    : `'${sheetTitle}'`

  const params = {
    auth,
    spreadsheetId,
    dateTimeRenderOption,
    valueRenderOption,
    majorDimension,
    range
  }

  // Permissive reading (ignore errors)
  try {
    const response = await fromCallback(cb => sheetsApi.spreadsheets.values.get(params, cb))
    let values = [...response.data.values]
    if (!config.readHeaders) {
      // Add generated headers:
      values.unshift(Array(values[0].length).fill(0).map((_, index) => toColumnName(index + 1)))
    }
    let headers = values[0]
    if (!jsonRaw) {
      values = values.map((row) => row.map((value, i) => csvFieldToJson(value, headers[i])))
    }
    return values
  } catch (ignore) {
    if (config.graceful) {
      log(`Sheet is unreadable. The document or the sheet may not exists yet. Original message: "${ignore.message}"`)
      return []
    } else {
      throw new Error(`Unable to read input - check input and permission (${ignore.message}) - Use option "graceful" to ignore`)
    }
  }
}

spreadstream.csvFieldToJson = function (value, key) {
  if (typeof value !== 'string') {
    return value
  }
  if (['true', 'false'].includes(value.toLowerCase())) {
    return value.toLowerCase() === 'true'
  } else if (value === '') {
    return null
  } else if (/^[-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?$/.test(value)) {
    return Number(value)
  }
  return value
}

/**
 * Generate a column name from column index
 * @param  {Number} num
 * @return {String} Column name (AB)
 * @author http://cwestblog.com/2013/09/05/javascript-snippet-convert-number-to-column-name/
 */
function toColumnName (num) {
  for (var ret = '', a = 1, b = 26; (num -= a) >= 0; a = b, b *= 26) {
    ret = String.fromCharCode(parseInt((num % b) / a) + 65) + ret
  }
  return ret
}
